File: platforms/android/app/src/main/java/com/ashera/capinsets/CapInsetsViewImpl.java
package com.ashera.capinsets;
import static com.ashera.widget.IWidget.BUFFER_STRATEGY_NONE;
import static com.ashera.widget.IWidget.UPDATE_UI_NONE;
import static com.ashera.widget.IWidget.UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE;

//start - imports
import java.util.*;

import android.content.Context;
import android.os.Build;
import android.view.View;
import android.annotation.SuppressLint;
import androidx.core.view.*;
import android.annotation.SuppressLint;

import com.ashera.widget.*;
import com.ashera.converter.*;
import android.widget.*;
import android.view.*;
import android.graphics.*;
import android.content.res.*;


import static com.ashera.widget.IWidget.*;
//end - imports
public class CapInsetsViewImpl implements com.ashera.widget.IAttributable {
	// start - body
	public final static String LOCAL_NAME = "CapInsetsView"; 
	private IWidget w;
	private CapInsetsViewImpl(IWidget widget) {
		this.w = widget;
	}
	
	public String getLocalName() {
		return LOCAL_NAME;
	}
	
	public CapInsetsViewImpl() {
	}
	
	@Override
	public com.ashera.widget.IAttributable newInstance(IWidget widget) {
		CapInsetsViewImpl newIntance = new CapInsetsViewImpl(widget);
		
		return newIntance;
	}
	
	
	@SuppressLint("NewApi")
	@Override
	public void loadAttributes(String localName) {

		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("backgroundCapInsets").withType("dimensiondppx").withOrder(-10));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("backgroundCapInsetsTop").withType("dimensiondppx").withOrder(-10));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("backgroundCapInsetsBottom").withType("dimensiondppx").withOrder(-10));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("backgroundCapInsetsLeft").withType("dimensiondppx").withOrder(-10));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("backgroundCapInsetsRight").withType("dimensiondppx").withOrder(-10));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("foregroundCapInsets").withType("dimensiondppx").withOrder(-10).withUiFlag(UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("foregroundCapInsetsTop").withType("dimensiondppx").withOrder(-10).withUiFlag(UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("foregroundCapInsetsBottom").withType("dimensiondppx").withOrder(-10).withUiFlag(UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("foregroundCapInsetsLeft").withType("dimensiondppx").withOrder(-10).withUiFlag(UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("foregroundCapInsetsRight").withType("dimensiondppx").withOrder(-10).withUiFlag(UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE));
	}

	@SuppressLint("NewApi")
	@Override
	public void setAttribute(WidgetAttribute key, String strValue, Object objValue, ILifeCycleDecorator decorator) {
		View view = (View) w.asWidget();
		switch (key.getAttributeName()) {
		case "backgroundCapInsets": {


		 applyCapInsets(w, objValue);



			}
			break;
		case "backgroundCapInsetsTop": {


		 applyCapInsetsTop(w, objValue);



			}
			break;
		case "backgroundCapInsetsBottom": {


		 applyCapInsetsBottom(w, objValue);



			}
			break;
		case "backgroundCapInsetsLeft": {


		 applyCapInsetsLeft(w, objValue);



			}
			break;
		case "backgroundCapInsetsRight": {


		 applyCapInsetsRight(w, objValue);



			}
			break;
		case "foregroundCapInsets": {


		 applyForegroundCapInsets(w, objValue);



			}
			break;
		case "foregroundCapInsetsTop": {


		 applyForegroundCapInsetsTop(w, objValue);



			}
			break;
		case "foregroundCapInsetsBottom": {


		 applyForegroundCapInsetsBottom(w, objValue);



			}
			break;
		case "foregroundCapInsetsLeft": {


		 applyForegroundCapInsetsLeft(w, objValue);



			}
			break;
		case "foregroundCapInsetsRight": {


		 applyForegroundCapInsetsRight(w, objValue);



			}
			break;
		default:
			break;
		}
	}
	
	@SuppressLint("NewApi")
	@Override
	public Object getAttribute(WidgetAttribute key, ILifeCycleDecorator decorator) {
		View view = (View) w.asWidget();
		switch (key.getAttributeName()) {
		}
		return null;
	}
	
	

	
private CapInsetsViewCommandBuilder builder;
private CapInsetsViewBean bean;
public CapInsetsViewBean getBean() {
	if (bean == null) {
		bean = new CapInsetsViewBean();
	}
	return bean;
}
public CapInsetsViewCommandBuilder getBuilder() {
	if (builder == null) {
		builder = new CapInsetsViewCommandBuilder();
	}
	return builder;
}


public  class CapInsetsViewCommandBuilder extends com.ashera.layout.ViewImpl.ViewCommandBuilder <CapInsetsViewCommandBuilder> {
    public CapInsetsViewCommandBuilder() {
	}
	
	public CapInsetsViewCommandBuilder execute(boolean setter) {
		if (setter) {
			w.executeCommand(command, null, IWidget.COMMAND_EXEC_SETTER_METHOD);
			w.getFragment().remeasure();
		}
		w.executeCommand(command, null, IWidget.COMMAND_EXEC_GETTER_METHOD);
return this;	}

public CapInsetsViewCommandBuilder setBackgroundCapInsets(String value) {
	Map<String, Object> attrs = initCommand("backgroundCapInsets");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setBackgroundCapInsetsTop(String value) {
	Map<String, Object> attrs = initCommand("backgroundCapInsetsTop");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setBackgroundCapInsetsBottom(String value) {
	Map<String, Object> attrs = initCommand("backgroundCapInsetsBottom");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setBackgroundCapInsetsLeft(String value) {
	Map<String, Object> attrs = initCommand("backgroundCapInsetsLeft");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setBackgroundCapInsetsRight(String value) {
	Map<String, Object> attrs = initCommand("backgroundCapInsetsRight");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setForegroundCapInsets(String value) {
	Map<String, Object> attrs = initCommand("foregroundCapInsets");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setForegroundCapInsetsTop(String value) {
	Map<String, Object> attrs = initCommand("foregroundCapInsetsTop");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setForegroundCapInsetsBottom(String value) {
	Map<String, Object> attrs = initCommand("foregroundCapInsetsBottom");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setForegroundCapInsetsLeft(String value) {
	Map<String, Object> attrs = initCommand("foregroundCapInsetsLeft");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CapInsetsViewCommandBuilder setForegroundCapInsetsRight(String value) {
	Map<String, Object> attrs = initCommand("foregroundCapInsetsRight");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
}
public class CapInsetsViewBean extends com.ashera.layout.ViewImpl.ViewBean{
		public CapInsetsViewBean() {
			super(CapInsetsViewImpl.this.w);
		}
public void setBackgroundCapInsets(String value) {
	getBuilder().reset().setBackgroundCapInsets(value).execute(true);
}

public void setBackgroundCapInsetsTop(String value) {
	getBuilder().reset().setBackgroundCapInsetsTop(value).execute(true);
}

public void setBackgroundCapInsetsBottom(String value) {
	getBuilder().reset().setBackgroundCapInsetsBottom(value).execute(true);
}

public void setBackgroundCapInsetsLeft(String value) {
	getBuilder().reset().setBackgroundCapInsetsLeft(value).execute(true);
}

public void setBackgroundCapInsetsRight(String value) {
	getBuilder().reset().setBackgroundCapInsetsRight(value).execute(true);
}

public void setForegroundCapInsets(String value) {
	getBuilder().reset().setForegroundCapInsets(value).execute(true);
}

public void setForegroundCapInsetsTop(String value) {
	getBuilder().reset().setForegroundCapInsetsTop(value).execute(true);
}

public void setForegroundCapInsetsBottom(String value) {
	getBuilder().reset().setForegroundCapInsetsBottom(value).execute(true);
}

public void setForegroundCapInsetsLeft(String value) {
	getBuilder().reset().setForegroundCapInsetsLeft(value).execute(true);
}

public void setForegroundCapInsetsRight(String value) {
	getBuilder().reset().setForegroundCapInsetsRight(value).execute(true);
}

}


	// end - body

//start - capinsets 
	private void applyForegroundCapInsets(IWidget w, Object objValue) {
		applyForegroundCapInsetsRight(w, objValue);
		applyForegroundCapInsetsLeft(w, objValue);
		applyForegroundCapInsetsTop(w, objValue);
		applyForegroundCapInsetsBottom(w, objValue);
	}
	private void applyForegroundCapInsetsRight(IWidget w, Object objValue) {
		String capInsetsKey = "capInsetsRight";
		if (isRTLayoutCapInsets(w)) {
			capInsetsKey = "capInsetsLeft";
		}
		w.applyAttributeCommand("foreground", "capInsets", new String[] {"foregroundCapInsetsTop", "foregroundCapInsetsBottom", "foregroundCapInsetsLeft", "foregroundCapInsetsRight"}, true, capInsetsKey, objValue);
	}

	private void applyForegroundCapInsetsLeft(IWidget w, Object objValue) {
		String capInsetsKey = "capInsetsLeft";
		if (isRTLayoutCapInsets(w)) {
			capInsetsKey = "capInsetsRight";
		}
		
		w.applyAttributeCommand("foreground", "capInsets", new String[] {"foregroundCapInsetsTop", "foregroundCapInsetsBottom", "foregroundCapInsetsLeft", "foregroundCapInsetsRight"}, true, capInsetsKey, objValue);
	}

	private void applyForegroundCapInsetsBottom(IWidget w, Object objValue) {
		w.applyAttributeCommand("foreground", "capInsets", new String[] {"foregroundCapInsetsTop", "foregroundCapInsetsBottom", "foregroundCapInsetsLeft", "foregroundCapInsetsRight"}, true, "capInsetsBottom", objValue);
	}

	private void applyForegroundCapInsetsTop(IWidget w, Object objValue) {
		w.applyAttributeCommand("foreground", "capInsets", new String[] {"foregroundCapInsetsTop", "foregroundCapInsetsBottom", "foregroundCapInsetsLeft", "foregroundCapInsetsRight"}, true, "capInsetsTop", objValue);
	}
	private void applyCapInsets(IWidget w, Object objValue) {
		applyCapInsetsRight(w, objValue);
		applyCapInsetsLeft(w, objValue);
		applyCapInsetsTop(w, objValue);
		applyCapInsetsBottom(w, objValue);
	}
	
	private void applyCapInsetsRight(IWidget w, Object objValue) {
		String capInsetsKey = "capInsetsRight";
		if (isRTLayoutCapInsets(w)) {
			capInsetsKey = "capInsetsLeft";
		}
		w.applyAttributeCommand("background", "capInsets", new String[] {"capInsetsTop", "capInsetsBottom", "capInsetsLeft", "capInsetsRight"}, true, capInsetsKey, objValue);
	}

	private void applyCapInsetsLeft(IWidget w, Object objValue) {
		String capInsetsKey = "capInsetsLeft";
		if (isRTLayoutCapInsets(w)) {
			capInsetsKey = "capInsetsRight";
		}
		
		w.applyAttributeCommand("background", "capInsets", new String[] {"capInsetsTop", "capInsetsBottom", "capInsetsLeft", "capInsetsRight"}, true, capInsetsKey, objValue);
	}

	private void applyCapInsetsBottom(IWidget w, Object objValue) {
		w.applyAttributeCommand("background", "capInsets", new String[] {"capInsetsTop", "capInsetsBottom", "capInsetsLeft", "capInsetsRight"}, true, "capInsetsBottom", objValue);
	}

	private void applyCapInsetsTop(IWidget w, Object objValue) {
		w.applyAttributeCommand("background", "capInsets", new String[] {"capInsetsTop", "capInsetsBottom", "capInsetsLeft", "capInsetsRight"}, true, "capInsetsTop", objValue);
	}

	private boolean isRTLayoutCapInsets(IWidget w) {
		if (com.ashera.layout.ViewImpl.isRTLLayoutDirection(w)) {
			return true;
		}
		
		View view  =(View) w.asWidget();
		if (!ViewCompat.isLayoutDirectionResolved(view)) {
			ViewGroup parent = (ViewGroup) view.getParent();
			while (parent != null) {
				if (ViewCompat.getLayoutDirection(parent) == View.LAYOUT_DIRECTION_RTL) {
					return true;
				}
				parent = (ViewGroup) parent.getParent();
			}
		}
		return false;
	}	
	//end - capinsets

}

File: platforms/android/app/src/main/tsc/src/android/widget/CapInsetsViewImpl.ts
// start - imports

	
import CommandAttr from '../../widget/CommandAttr';
import IWidget from '../../widget/IWidget';
import ILayoutParam from '../../widget/ILayoutParam';
import {plainToClass, Type, Exclude, Expose, Transform} from "class-transformer";
import 'babel-polyfill';
import {Gravity} from '../../widget/TypeConstants';
import {ITranform, TransformerFactory} from '../../widget/TransformerFactory';
import {Event} from '../../app/Event';
import {MotionEvent} from '../../app/MotionEvent';
import {DragEvent} from '../../app/DragEvent';
import {KeyEvent} from '../../app/KeyEvent';
import { ScopedObject } from '../../app/ScopedObject';













// end - imports
import {ViewImpl} from './ViewImpl';
export abstract class CapInsetsViewImpl<T> extends ViewImpl<T>{
	//start - body
	static initialize() {
    }	
	@Type(() => CommandAttr)
	@Expose({ name: "backgroundCapInsets" })
	backgroundCapInsets!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "backgroundCapInsetsTop" })
	backgroundCapInsetsTop!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "backgroundCapInsetsBottom" })
	backgroundCapInsetsBottom!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "backgroundCapInsetsLeft" })
	backgroundCapInsetsLeft!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "backgroundCapInsetsRight" })
	backgroundCapInsetsRight!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "foregroundCapInsets" })
	foregroundCapInsets!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "foregroundCapInsetsTop" })
	foregroundCapInsetsTop!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "foregroundCapInsetsBottom" })
	foregroundCapInsetsBottom!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "foregroundCapInsetsLeft" })
	foregroundCapInsetsLeft!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "foregroundCapInsetsRight" })
	foregroundCapInsetsRight!:CommandAttr<string>| undefined;

	@Exclude()
	protected thisPointer: T;	
	protected abstract getThisPointer(): T;
	reset() : T {	
		super.reset();
		this.backgroundCapInsets = undefined;
		this.backgroundCapInsetsTop = undefined;
		this.backgroundCapInsetsBottom = undefined;
		this.backgroundCapInsetsLeft = undefined;
		this.backgroundCapInsetsRight = undefined;
		this.foregroundCapInsets = undefined;
		this.foregroundCapInsetsTop = undefined;
		this.foregroundCapInsetsBottom = undefined;
		this.foregroundCapInsetsLeft = undefined;
		this.foregroundCapInsetsRight = undefined;
		return this.thisPointer;
	}
	constructor(id: string, path: string[], event:  string) {
		super(id, path, event);
		this.thisPointer = this.getThisPointer();
	}
	

	public setBackgroundCapInsets(value : string) : T {
		this.resetIfRequired();
		if (this.backgroundCapInsets == null || this.backgroundCapInsets == undefined) {
			this.backgroundCapInsets = new CommandAttr<string>();
		}
		
		this.backgroundCapInsets.setSetter(true);
		this.backgroundCapInsets.setValue(value);
		this.orderSet++;
		this.backgroundCapInsets.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setBackgroundCapInsetsTop(value : string) : T {
		this.resetIfRequired();
		if (this.backgroundCapInsetsTop == null || this.backgroundCapInsetsTop == undefined) {
			this.backgroundCapInsetsTop = new CommandAttr<string>();
		}
		
		this.backgroundCapInsetsTop.setSetter(true);
		this.backgroundCapInsetsTop.setValue(value);
		this.orderSet++;
		this.backgroundCapInsetsTop.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setBackgroundCapInsetsBottom(value : string) : T {
		this.resetIfRequired();
		if (this.backgroundCapInsetsBottom == null || this.backgroundCapInsetsBottom == undefined) {
			this.backgroundCapInsetsBottom = new CommandAttr<string>();
		}
		
		this.backgroundCapInsetsBottom.setSetter(true);
		this.backgroundCapInsetsBottom.setValue(value);
		this.orderSet++;
		this.backgroundCapInsetsBottom.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setBackgroundCapInsetsLeft(value : string) : T {
		this.resetIfRequired();
		if (this.backgroundCapInsetsLeft == null || this.backgroundCapInsetsLeft == undefined) {
			this.backgroundCapInsetsLeft = new CommandAttr<string>();
		}
		
		this.backgroundCapInsetsLeft.setSetter(true);
		this.backgroundCapInsetsLeft.setValue(value);
		this.orderSet++;
		this.backgroundCapInsetsLeft.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setBackgroundCapInsetsRight(value : string) : T {
		this.resetIfRequired();
		if (this.backgroundCapInsetsRight == null || this.backgroundCapInsetsRight == undefined) {
			this.backgroundCapInsetsRight = new CommandAttr<string>();
		}
		
		this.backgroundCapInsetsRight.setSetter(true);
		this.backgroundCapInsetsRight.setValue(value);
		this.orderSet++;
		this.backgroundCapInsetsRight.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setForegroundCapInsets(value : string) : T {
		this.resetIfRequired();
		if (this.foregroundCapInsets == null || this.foregroundCapInsets == undefined) {
			this.foregroundCapInsets = new CommandAttr<string>();
		}
		
		this.foregroundCapInsets.setSetter(true);
		this.foregroundCapInsets.setValue(value);
		this.orderSet++;
		this.foregroundCapInsets.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setForegroundCapInsetsTop(value : string) : T {
		this.resetIfRequired();
		if (this.foregroundCapInsetsTop == null || this.foregroundCapInsetsTop == undefined) {
			this.foregroundCapInsetsTop = new CommandAttr<string>();
		}
		
		this.foregroundCapInsetsTop.setSetter(true);
		this.foregroundCapInsetsTop.setValue(value);
		this.orderSet++;
		this.foregroundCapInsetsTop.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setForegroundCapInsetsBottom(value : string) : T {
		this.resetIfRequired();
		if (this.foregroundCapInsetsBottom == null || this.foregroundCapInsetsBottom == undefined) {
			this.foregroundCapInsetsBottom = new CommandAttr<string>();
		}
		
		this.foregroundCapInsetsBottom.setSetter(true);
		this.foregroundCapInsetsBottom.setValue(value);
		this.orderSet++;
		this.foregroundCapInsetsBottom.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setForegroundCapInsetsLeft(value : string) : T {
		this.resetIfRequired();
		if (this.foregroundCapInsetsLeft == null || this.foregroundCapInsetsLeft == undefined) {
			this.foregroundCapInsetsLeft = new CommandAttr<string>();
		}
		
		this.foregroundCapInsetsLeft.setSetter(true);
		this.foregroundCapInsetsLeft.setValue(value);
		this.orderSet++;
		this.foregroundCapInsetsLeft.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setForegroundCapInsetsRight(value : string) : T {
		this.resetIfRequired();
		if (this.foregroundCapInsetsRight == null || this.foregroundCapInsetsRight == undefined) {
			this.foregroundCapInsetsRight = new CommandAttr<string>();
		}
		
		this.foregroundCapInsetsRight.setSetter(true);
		this.foregroundCapInsetsRight.setValue(value);
		this.orderSet++;
		this.foregroundCapInsetsRight.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		
	//end - body

}
	
//start - staticinit

export class CapInsetsView extends CapInsetsViewImpl<CapInsetsView> implements IWidget{
    getThisPointer(): CapInsetsView {
        return this;
    }
    
   	public getClass() {
		return CapInsetsView;
	}
	
   	constructor(id: string, path: string[], event: string) {
		super(id, path, event);	
	}
}

CapInsetsViewImpl.initialize();

//end - staticinit
